chainer/utils/sparse.py

```diff
@@ -0,0 +1,102 @@
+import chainer
+from chainer.backends import cuda
+
+
+class CooMatrix(object):
+    """A sparse matrix in COO format.
+
+    Args:
+        data (numpy.ndarray or cupy.ndarray): The entries of the matrix.
+            The entries are usually non-zero-elements in the matrix.
+        row (numpy.ndarray or cupy.ndarray): The row indices of the matrix
+            entries.
+        col (numpy.ndarray or cupy.ndarray): The column indices of the matrix
+            entries.
+        shape (tuple of int): The shape of the matrix in dense format.
+        requires_grad (bool): If ``True``, gradient of this sparse matrix will
+            be computed in back-propagation.
+    """
+
+    def __init__(self, data, row, col, shape, requires_grad=False):
+        if not (1 <= data.ndim <= 2):
+            raise ValueError('ndim of data must be 1 or 2.')
+        if not (data.ndim == row.ndim == col.ndim):
+            raise ValueError('ndim of data, row and col must be the same.')
+        if len(shape) != 2:
+            raise ValueError('length of shape must be 2.')
+        if not (shape[0] > 0 and shape[1] > 0):
+            raise ValueError('numbers in shape must be greater than 0.')
+        self.data = chainer.Variable(data, requires_grad=requires_grad)
+        self.row = row
+        self.col = col
+        self.shape = shape  # (row, col)
+
+    def to_dense(self):
+        """Returns a dense matrix format of this sparse matrix."""
+        data = self.data.data
+        xp = cuda.get_array_module(data)
+        if data.ndim == 1:
+            x = xp.zeros(self.shape, dtype=data.dtype)
+            x[self.row, self.col] = data
+            return x
+        if data.ndim == 2:
+            nb = data.shape[0]
+            x = xp.zeros((nb, self.shape[0], self.shape[1]),
+                         dtype=data.dtype)
+            for i in range(nb):
+                nnz = xp.count_nonzero(data[i])
+                x[i, self.row[i, :nnz],
+                  self.col[i, :nnz]] = data[i, :nnz]
+            return x
+
+
+def to_coo(x, ldnz=None, requires_grad=False):
+    """Returns a sparse matrix in COO format of a dense matrix.
+
+    Args:
+        x (numpy.ndarray or cupy.ndarray): Input dense matrix. The ndim of
+            ``x`` must be two or three. If ndim is two, it is treated as
+            a single matrix. If three, it is treated as batched matrices.
+        ldnz (int): Size of arrays for data, row index and column index to be
+            created. The Actual size becomes max(nnz, ldnz) where nnz is number
+            of non-zero elmeents in a input dense matrix.
```

Comment:
Oh, a small typo at `elemeents`.

Reply:
Just a small question, is "ldnz" common terminology, would you mind explaining what is stands for?

Reply:
Thanks, will fix the typo.

Re: "ldnz", it stands for "Leading Dimension of array for Non-Zero elements" and shows the size of array which holds matrix entries, row indexes or column indexes. I follow naming convention of matrix libraries like blas, lapck, etc. in which, for example, scalar variable `ldA` indicates the size of array `A`.

Reply:
Thanks you for your explanation! I see it is used for indexing in general. It seems though that it is not tested. Should we add tests since it's a public interface?

Reply:
Thank you so much! I found a bug in `to_dense` of `CooMatrix` thanks to the tests for `ldnz`.